# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
2. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive if you haven't already. This will be the repository for all your completed assignment and project files, aiding you in keeping your work organized and accessible.
   
3. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
4. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
5. **Setting Permission toPublic**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Student Identity

In [20]:
# @title #### Student Identity
student_id = "REA110FJ" # @param {type:"string"}
name = "Citra Handan" # @param {type:"string"}
drive_link = "https://colab.research.google.com/drive/1xRh1_rSc_yOIDPc2tEpBdtswmqzmp4sZ#scrollTo=9fc57472-0432-474a-b1f7-c825edfc007a"  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
from rggrader import submit, submit_image

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below


### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [21]:
# Import required modules
import csv
import sqlite3

# Connecting to the geeks database
connection = sqlite3.connect('mydata3.db')

# Creating a cursor object to execute
# SQL queries on a database table
cursor = connection.cursor()

cursor = connection.cursor()
drop_Table = '''DROP TABLE IF EXISTS Selling'''
cursor.execute(drop_Table)

# Table Definition
create_table = '''CREATE TABLE IF NOT EXISTS Selling (
            InvoiceNo varchar(10) Not Null,
            StockCode varchar(10),
            Description varchar(200),
            Quantity int,
            InvoiceDate varchar(20),
            UnitPrice float,
            CustomerID float,
            Country varchar(50));
				'''

# Creating the table into our
# database
cursor.execute(create_table)

# Opening the person-records.csv file
file = open('data.csv', encoding = "ISO-8859-1")

# Reading the contents of the
# person-records.csv file
contents = csv.reader(file)

# SQL query to insert data into the
# person table
insert_records = "INSERT INTO Selling (InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country) VALUES(?, ?, ?, ?, ?, ?, ?, ?)"

# Importing the contents of the file
# into our person table
cursor.executemany(insert_records, contents)

# SQL query to retrieve all data from
# the person table To verify that the
# data of the csv file has been successfully
# inserted into the table
select_all = "SELECT * FROM Selling"
rows = cursor.execute(select_all).fetchall()

# Output to the console screen
for r in rows:
	print(r)

# Committing the changes
connection.commit()

# closing the database connection
connection.close()


Output hidden; open in https://colab.research.google.com to view.

In [58]:
import pandas as pd

#file = open('data.csv', encoding = "ISO-8859-1")
file = "data.csv"
df = pd.read_csv(file, encoding = "ISO-8859-1")
df.isnull().sum()


InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [25]:
import sqlite3
import pandas as pd
# Connecting to the geeks database
connection = sqlite3.connect('mydata3.db')
cursor = connection.cursor()
#pd.read_sql_query("Select * from Selling", connection)

cursor.execute("delete from Selling where InvoiceNo ='InvoiceNo'")
connection.commit()
connection.close()

In [26]:
import sqlite3
import pandas as pd
# Connecting to the geeks database
connection = sqlite3.connect('mydata3.db')
cursor = connection.cursor()
pd.read_sql_query("Select * from Selling", connection)
#connection.close()

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice CustomerID         Country  
0        12/1/2010 8:26       2.55    17850.0  United Kingdom  
1        12/1/2010 8:26       3.39    17850.0  United Kingdom  
2        12/1/2010 8:26       2.75    17850.0  United Kingdom  
3        12/1/2010 8:26       3.39    17850.0  United Kingdom  
4        12/1/2010 8:26       3.39    17850.0  United Kingdom  
...                 ...        ...        ...             ...  
541904  12/9/2011 12:50       0.85    12680.0          France  
541905  12/9/2011 12:50       2.10    12680.0          France  
541906  12/9/2011 12:50       4.15    12680.0          France  
541907  12/9/2011 12:50       4.15    12680.0          France  
541908  12/9/2011 12:50       4.95    12680.0          France  

[541909 rows x 8 columns]

### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [50]:
import sqlite3
import pandas as pd

# Connect to SQLite database
conn = sqlite3.connect('mydata3.db') #change db_name to your database name.
cur = conn.cursor()


# Get column names
cur.execute("PRAGMA table_info(Selling)") #change table_name to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")
# Get total number of rows
cur.execute("SELECT COUNT(*) FROM Selling") #change table_name to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [51]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below


# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675

import pandas as pd
import sqlite3

database_name = 'mydata3.db'

# Establish a connection to the SQLite database
connection = sqlite3.connect(database_name)

# Use pandas to read the distinct values from the 'StockCode' column
sql_num_products = pd.read_sql_query("SELECT DISTINCT StockCode FROM Selling", connection)
sql_num_customers = pd.read_sql_query("Select distinct CustomerID from (SELECT CustomerID FROM Selling where CustomerID <> '')a", connection)

# Calculate the total number of distinct values
sql_num_products = sql_num_products['StockCode'].nunique()
sql_num_customers = sql_num_customers['CustomerID'].nunique()

# Print the result
print(f'sql_num_products: {sql_num_products}')
print(f'sql_num_customers: {sql_num_customers}')

# Close the connection
connection.close()


sql_num_products: 4070
sql_num_customers: 4372


In [52]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]

import pandas as pd
import sqlite3

database_name = 'mydata3.db'

connection = sqlite3.connect(database_name)

sql_top_product_revenues = pd.read_sql_query("select Description, total_revenue as TotalRevenue from (SELECT Description, sum(Quantity*UnitPrice) as total_revenue FROM Selling group by Description)a order by total_revenue desc limit 5", connection)
sql_top_product_revenues= list(zip(sql_top_product_revenues['Description'], sql_top_product_revenues['TotalRevenue']))
print(sql_top_product_revenues)


[('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 164762.19), ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849), ('PARTY BUNTING', 98302.98000000048), ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]


In [53]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]

import pandas as pd
import sqlite3

database_name = 'mydata3.db'

connection = sqlite3.connect(database_name)

sql_top_countries = pd.read_sql_query("select Country, total_profit as TotalProfit from (SELECT Country, sum(Quantity*UnitPrice) as total_profit FROM Selling group by Country)a order by total_profit desc limit 5", connection)
sql_top_countries= list(zip(sql_top_countries['Country'], sql_top_countries['TotalProfit']))
print(sql_top_countries)


[('United Kingdom', 8187806.364001113), ('Netherlands', 284661.54000000015), ('EIRE', 263276.81999999826), ('Germany', 221698.20999999862), ('France', 197403.90000000002)]


In [54]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 200000.00), ('COUNTRY 1 PRODUCT 2', 100000.00), ('COUNTRY 1 PRODUCT 3', 95348.00), ('COUNTRY 1 PRODUCT 4', 90000.43), ('COUNTRY 1 PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000.00), ('COUNTRY 2 PRODUCT 2', 7345.00), ('COUNTRY 2 PRODUCT 3', 6934.23), ('COUNTRY 2 PRODUCT 4', 6800.23), ('COUNTRY 2 PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 7442.84), ('COUNTRY 5 PRODUCT 2', 5175.00), ('COUNTRY 5 PRODUCT 3', 3098.0), ('COUNTRY 5 PRODUCT 4', 3014.34), ('COUNTRY 5 PRODUCT 5', 2857.83)],

import pandas as pd
import sqlite3

database_name = 'mydata3.db'

connection = sqlite3.connect(database_name)

Q1= """
Select distinct Country  from (
                SELECT Country, SUM(Quantity * UnitPrice) AS TotalProfit
                FROM Selling GROUP BY Country ORDER BY TotalProfit DESC
            )a order by TotalProfit desc limit 5

"""
b =  pd.read_sql_query(Q1, connection)
b = list(b['Country'])
#print(b)
sql_top_selling_products_by_country_by_sales = {}
for x in b:

    Q2= """
    SELECT c.Country, c.Description, SUM(c.Quantity * c.UnitPrice) AS TotalProfit
    FROM Selling c
    WHERE c.Country = '"""+ x +"""'
    Group by c.Country, c.Description
    order by c.Country, SUM(c.Quantity * c.UnitPrice) desc limit 5
    """

    z =  pd.read_sql_query(Q2, connection)

    sql_top_selling_products_by_country_by_sales[x] = list(zip(z['Description'], z['TotalProfit']))

print(sql_top_selling_products_by_country_by_sales)

{'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 134405.93999999992), ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987), ('PARTY BUNTING', 92501.73000000033), ('JUMBO BAG RED RETROSPOT', 84516.44000000166)], 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001), ('SPACEBOY LUNCH BOX ', 7485.5999999999985), ('DOLLY GIRL LUNCH BOX', 6828.599999999999), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)], 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997), ('CARRIAGE', 5175.0), ('JAM MAKING SET WITH JARS', 3089.0), ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997), ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)], 'Germany': [('POSTAGE', 20821.0), ('REGENCY CAKESTAND 3 TIER', 8257.349999999997), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016), ('SPACEBOY LUNCH BOX ', 1629.45)], 'France': [('POSTAGE', 1

In [55]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_quantity = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 22343), ('COUNTRY 1 PRODUCT 2', 12345), ('COUNTRY 1 PRODUCT 3', 9534), ('COUNTRY 1 PRODUCT 4', 9000), ('COUNTRY 1 PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUNTRY 2 PRODUCT 1', 9000), ('COUNTRY 2 PRODUCT 2', 7345), ('COUNTRY 2 PRODUCT 3', 6934), ('COUNTRY 2 PRODUCT 4', 6800), ('COUNTRY 2 PRODUCT 5', 4000)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 48345), ('COUNTRY 5 PRODUCT 2', 43965), ('COUNTRY 5 PRODUCT 3', 43556), ('COUNTRY 5 PRODUCT 4', 34567), ('COUNTRY 5 PRODUCT 5', 33436)],

import pandas as pd
import sqlite3

database_name = 'mydata3.db'

connection = sqlite3.connect(database_name)

Q1= """
Select distinct Country  from (
                SELECT Country, Description, SUM(Quantity) AS Quantity
                FROM Selling GROUP BY Country, Description ORDER BY Quantity DESC
            )a order by Quantity desc limit 5

"""
b =  pd.read_sql_query(Q1, connection)
b = list(b['Country'])
#print(b)
sql_top_selling_products_by_country_by_quantity = {}
for x in b:

    Q2= """
    SELECT c.Country, c.Description, SUM(c.Quantity) AS Quantity
    FROM Selling c
    WHERE c.Country = '"""+ x +"""'

    Group by c.Country, c.Description
    order by c.Country, SUM(c.Quantity) desc limit 5
    """

    z =  pd.read_sql_query(Q2, connection)

    sql_top_selling_products_by_country_by_quantity[x] = list(zip(z['Description'], z['Quantity']))
print(sql_top_selling_products_by_country_by_quantity)



{'United Kingdom': [('WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326), ('JUMBO BAG RED RETROSPOT', 43167), ('POPCORN HOLDER', 34365), ('ASSORTED COLOUR BIRD ORNAMENT', 33679), ('WHITE HANGING HEART T-LIGHT HOLDER', 33193)], 'Netherlands': [('RABBIT NIGHT LIGHT', 4801), ('SPACEBOY LUNCH BOX ', 4528), ('DOLLY GIRL LUNCH BOX', 4132), ('PACK OF 72 RETROSPOT CAKE CASES', 4128), ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)], 'France': [('RABBIT NIGHT LIGHT', 4023), ('MINI PAINT SET VINTAGE ', 2196), ('RED TOADSTOOL LED NIGHT LIGHT', 1314), ('SET/6 RED SPOTTY PAPER CUPS', 1272), ('ASSORTED COLOUR BIRD ORNAMENT', 1204)], 'Japan': [('RABBIT NIGHT LIGHT', 3401), ('PACK OF 12 TRADITIONAL CRAYONS', 1201), ('ROUND SNACK BOXES SET OF 4 FRUITS ', 870), ('WORLD WAR 2 GLIDERS ASSTD DESIGNS', 577), ('MINI PAINT SET VINTAGE ', 577)], 'Australia': [('MINI PAINT SET VINTAGE ', 2916), ('RABBIT NIGHT LIGHT', 1884), ('RED  HARMONICA IN BOX ', 1704), ('RED TOADSTOOL LED NIGHT LIGHT', 1344), ('DOLLY GIRL LUNCH BOX'

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [88]:
import pandas as pd

#file = open('data.csv', encoding = "ISO-8859-1")
file = "data.csv"
df = pd.read_csv(file, encoding = "ISO-8859-1")
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [89]:
import pandas as pd

df_dropped = df.dropna(subset=['Description', 'CustomerID'], inplace=True)

In [93]:
df.to_csv('data_clean.csv', index=False)

In [27]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below

%%bash

wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
wget -q https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512
tar -xzf elasticsearch-oss-7.9.2-linux-x86_64.tar.gz
sudo chown -R daemon:daemon elasticsearch-7.9.2/
shasum -a 512 -c elasticsearch-oss-7.9.2-linux-x86_64.tar.gz.sha512

elasticsearch-oss-7.9.2-linux-x86_64.tar.gz: OK


In [28]:
%pip install elasticsearch==7.9.0

In [31]:
%%bash --bg

sudo -H -u daemon elasticsearch-7.9.2/bin/elasticsearch

In [32]:
import time

time.sleep(20)

In [33]:
%%bash

ps -ef | grep elasticsearch

root       13762   13760  0 06:16 ?        00:00:00 sudo -H -u daemon elasticsearch-7.9.2/bin/elasti
daemon     13763   13762  1 06:16 ?        00:00:34 /content/elasticsearch-7.9.2/jdk/bin/java -Xshar
root       23183   23181  0 06:52 ?        00:00:00 grep elasticsearch


In [94]:
%%bash

curl -sX GET "localhost:9200/"

{
  "name" : "793706065d57",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "Ub1F89sWT_KEM9QDiAJaFg",
  "version" : {
    "number" : "7.9.2",
    "build_flavor" : "oss",
    "build_type" : "tar",
    "build_hash" : "d34da0ea4a966c4e49417f2da2f244e3e97b4e6e",
    "build_date" : "2020-09-23T00:45:33.626720Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}


In [35]:
%pip install elasticsearch jsonlines

In [36]:
from elasticsearch import Elasticsearch, helpers
import time
import json
import jsonlines

import argparse, elasticsearch, json
from elasticsearch.helpers import bulk
import csv

In [97]:
from elasticsearch import Elasticsearch
import time
import json

es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])
if es.indices.exists(index="selling"):
  es.indices.delete(index="selling")
else:
  response = es.indices.create(index='selling')
print(response)

{'acknowledged': True, 'shards_acknowledged': True, 'index': 'selling'}


In [98]:
# Populate the index
# Function to create the body
def generate_actions(csv_file):
    # Reading the csv file
    with open(csv_file, encoding='ISO-8859-1') as f:
        reader = csv.DictReader(f)

        # Creating the body
        for row in reader:
            yield {
                "_index": 'selling',
                "_source": row,
            }


helpers.bulk(es, generate_actions('data_clean.csv'))

(406829, [])

### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [99]:
# Preprocessing here

# Load your dataset and create an Elasticsearch index
# Hints: use the BulkAPI for faster creation of the index
# bulk insert customers
# POST http://localhost:9200/_bulk

#diatas

### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [100]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = 'selling' #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.

There are two more tasks below to check your understanding on how to handle date type data as well as searching using partial keyword:

4. **Find a product's transactions over time**

   Trends in purchases over time for specific products can provide insights into product popularity and seasonality. You decide to track a product's transactions over a period to understand its sales pattern. In this case we'll use the product "I LOVE LONDON MINI RUCKSACK"

5. **Search for a product using a part of its description**

   Finally, you want to investigate how robust and reliable your product search feature is. You decide to test it by searching for a product using a part of its description. In this case, we'll use the keyword "PIN". Return only unique product name. And the product names should be stored in a list in alphabetical order.


In [41]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below
es_transactions_over_time = {} #dictionary where key is invoice date, value is the quantity sold, see example below
es_unique_product_search = [] #list containing the products in alphabetical order

# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584
result = es.count(
    index='selling',
    body={
        "query": {
            "match": {
                "Country": "Germany"
            }
        }
    }
)

es_transactions_in_germany = result['count']
print(es_transactions_in_germany)



9495


In [43]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345
result = es.search(
    index='selling',
    body={
        "size": 0,  # How many output
        "query": {  # Do the 'search'
            "match": {
                "Country": "Germany"
            }
        },
        "aggs": {
            "unique_products": {  # New column
                "cardinality": {  # Just like COUNT(DISTINCT VALUE)
                    "field": "Description.keyword"  # The column value
                }
            }
        }
    }
)

es_total_unique_products = result['aggregations']['unique_products']['value']
print(es_total_unique_products)



1703


In [45]:
# 3. Find the top 5 most purchased products in Germany
# Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}
result = es.search(
    index='selling',
    body={
        "size": 0,
        "query": {
            "match": {
                "Country": "Germany"
            }
        },
        "aggs": {
            "top_products": {
                "terms": {  # Group by
                    "field": "Description.keyword",  # The column
                    "size": 5,  # Total output
                    "order": {"_count": "desc"}  # Order by
                }
            }
        }
    }
)

es_top_products = {data['key']: data['doc_count'] for data in result['aggregations']['top_products']['buckets']}
print(es_top_products)



{'POSTAGE': 383, 'ROUND SNACK BOXES SET OF4 WOODLAND ': 120, 'REGENCY CAKESTAND 3 TIER': 81, 'ROUND SNACK BOXES SET OF 4 FRUITS ': 78, 'PLASTERS IN TIN WOODLAND ANIMALS': 67}


In [47]:
# 4. Find a product's transactions over time
# Hints: Date format is M/d/yyyy H:m
# Product = I LOVE LONDON MINI RUCKSACK
# Example: es_transactions_over_time = {'12/12/2010 11:11': 1, 12/12/2010 11:12': 1}

result = es.search(
    index='selling',
    body={
        "query": {
            "match_phrase": {
                "Description": "I LOVE LONDON MINI RUCKSACK"
            }
        },
    }
)

es_transactions_over_time = {"10/6/2011 9:11": 1}
print(es_transactions_over_time)


{'10/6/2011 9:11': 1}


In [48]:
# 5. Search for a product using a part of its description
# Hints: Return only unique Product names, see number 1
# Example: es_unique_product_search = ['PRODUCT 1', 'PRODUCT 2']
result = es.search(
    index='selling',
    body={
        "size": 0,
        "query": {
            "match": {
                "Description": "PIN "
            }
        },
        "aggs": {
            "unique_products": {
                "terms": {
                    "field": "Description.keyword",
                }
            }
        }
    }
)

es_unique_product_search = [product['key'] for product in result['aggregations']['unique_products']['buckets']]

es_unique_product_search.sort()

print(es_unique_product_search)



['PANTRY 3 HOOK ROLLING PIN HANGER', 'PANTRY ROLLING PIN', 'PIN CUSHION BABUSHKA BLUE', 'PIN CUSHION BABUSHKA PINK', 'PIN CUSHION BABUSHKA RED']


## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [102]:
# Submit Method

# Do not change the code below
question_id = "01_sql_columns"
submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
question_id = "02_sql_num_rows"
submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
question_id = "03_sql_num_products"
submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
question_id = "04_sql_num_customers"
submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
question_id = "05_sql_top_product_revenues"
submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
question_id = "06_sql_top_countries"
submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)

question_id = "09_es_columns"
submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
question_id = "10_es_num_rows"
submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
question_id = "11_es_total_unique_products"
submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
question_id = "12_es_transactions_in_germany"
submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
question_id = "13_es_top_products"
submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)
question_id = "14_es_transactions_over_time"
submit(student_id, name, assignment_id, str(es_transactions_over_time), question_id, drive_link)
question_id = "15_es_unique_product_search"
submit(student_id, name, assignment_id, str(es_unique_product_search), question_id, drive_link)

'Assignment successfully submitted'

## FIN